---
title: "Introduction"
description: |
  In this section ...
#author:
#  - name: "Maggie D Johnson"
#    url: https://github.com/jjallaire
#    affiliation: STRI
#    affiliation_url: https://www.rstudio.com
#  - name: "Noelle Lucey"
#    url: https://github.com/yihui
#    affiliation: STRI
#    affiliation_url: https://www.rstudio.com
bibliography: assets/cite.bib
---

```{r setup, include=FALSE}
remove(list = ls())
library(dplyr)
library(DT)
library(ggplot2)
library(scales)
library(magick)
library(gridExtra)
library(grid)

```

In this study, we looked at INSERT DETAIL</span>.

# Goals of the Study


# Definitions & Abbreviations
* Amplicon Sequence Variant (**ASV**): Exact sequence variant---analogous to an OTU---but with single nucleotide resolution.
* Differentially abundant (**DA**) feature: Taxa, ASV, etc. that is disproportionately abundant in a group of samples and statistically different than other groups.

# Workflow

You can access each step of the workflow by using the navigation bar at the top of the page. Below is a brief description of each step.

## A. Field Analyses

üëã Hi Maggie & Noelle!!!! I bet you really, really want to add content to this section ‚ò∫Ô∏è.

## B. 16S rRNA Analysis

We sequence a total of 30 

### No 1. DADA2 Workflow

In this part we go through the steps of processing raw 16S rRNA read data including assessing read quality, filtering reads, correcting errors, and infersing amplicon sequence variants (ASVs).

### No 2. Data Preparation

Next we go through the steps of defining sample groups, creating phyloseq objects, removing unwanted samples, and removing contaminant ASVs. Various parts of this section can easily be modified to perform different analyses. For example, if you were only interested in a specific taxa or group of samples, you could change the code here to create new phyloseq objects.

### No 3. Water Sample Diversity

Now we move are attention to just the water samples. We wanted to understand how ASVs partitioned between normoxic and hypoxic conditions. We also wanted to assess the specificity of each ASV to determine habitat preference. To our knowledge there is no quantitative way to do this. The only attempt we are aware of was [MetaMetaDB](http://mmdb.aori.u-tokyo.ac.jp/) but it is based on a 454 database and no longer seems to be in active development. So we used an approach based on the work of [Sullam *et. al.*](https://doi.org/10.1111/j.1365-294X.2012.05552.x), first identifying differentially abundant ASVs, then searching for closest database hits, and finally using phylogenetic analysis and top hit metadata (isolation source, natural host) to infer habitat preference.

## C. Water Column Metagenomes

We sequenced four samples for metagenomic analysis--two from Crawl Caye and two from Cayo Roldan. The Crawl Caye


### No 5. Setup Working Environment</a>
For the metagenomic analysis we largely leave the R environment and enter the [anvi'o](http://merenlab.org/software/anvio/) ecosystem. The first major step in our analysis of  metagenomic samples from hypoxic and normoxic conditions is to setup our working environment. We will do this in two steps. The first is to install the backbone of our computational operations, which consists of Miniconda and anvio. Next we build the annotation databases and install the associated tools. We provide complete details of these steps.


### No 6. Annotation Databases</a>

Next we describe the various database we used to annotate the metagenomes. There are two main types of annotations we are interested in for this metagenomic project‚Äîtaxonomic and functional‚Äîand there are many, many ways to accomplish both of these goals. This next section involves building the databases and installing any additional tools we need for annotation.


### No 7. Assembly & Annotations</a>

Here we move to the processing part of the workflow, which involves the following: adapter trimming of raw data; quality filtering of trimmed reads; co-assembling  QCed reads; mapping reads to the assembly; profiling the mapping results; merging profile dbs; classify reads and genes; annotating genes; and running HMM profiles.


### No 8. Assembley & Annotation Summary</a>

In this section of the workflow we summarize the QC & assemby results, the Kraken short read taxonomy, mapping results. and contig classifications.

# Publication Related Material

## Figures & Tables

Use the links below if you want to jump directly to the code used to produce the **figures and tables** from the original publication. You can also find the full Supplementary files for the paper  [here](supplemental-material.html) but there is no R code on this page. If you want to see the code that produced the supplemental material, the direct links are also below. 

### Main Paper

INSERT LINKS TO MAIN FIGURES AND TABLES

### Supplemental Material

INSERT LINKS TO SUPP FIGURES AND TABLES

# Color & Graphics

We will rely on color a lot to display information about diversity, distinguish between sample types, etc. Many of us have different abilities to perceive color and/or differences in color. [Color blindness](http://mkweb.bcgsc.ca/colorblind/index.mhtml#page-container), also known as color vision deficiency, is a decreased ability to see color or differences in color. It is important that we use **a**) a relatively small color palette and **b**) a palette that is friendly to a variety of people.

Bang Wong has a paper in Nature Methods paper called [Points of view: Color blindness](http://dx.doi.org/10.1038/nmeth.1618)[@wong2011points] where proposes a color-blind friendly color palette that uses contrasting colors that can be distinguished by a range of people. Consider that roughly 8% of people (mostly males) are color blind. So what do you think? Do you want Keanu Reeves to understand your figures or not?

Wong‚Äôs scheme is conservative‚Äî--there are only 8 colors.  Martin Krzywinsk has a marvelous post on [designing color friendly palettes](http://mkweb.bcgsc.ca/colorblind/palettes.mhtml#page-container) that includes additional [12, 15, and 24](http://mkweb.bcgsc.ca/colorblind/img/colorblindness.palettes.v11.pdf) color palette schemes, which offer more options than Wong‚Äôs 8-color palette. We just need to be careful‚Äî‚Äìfigures with too many colors can inhibit our ability to discern patterns. Limited color palettes like these force us to be more selective when deciding what information to target or how many groups to display. 

## Creating Color Palletes

Here are a few color palette examples. A light grey  (`#7F7F7F`) was added to the 8-color palette. Shown here are the hex codes for these palettes but if you want the color names and [sRGB](https://en.wikipedia.org/wiki/RGB_color_model) codes you can download plain text versions of the palettes from Martin's website [here](http://mkweb.bcgsc.ca/colorblind/palettes/color.blindness.palettes.v11.tgz).

If you are interested in the R code used to make these palettes, click the arrow to expand the code block. 

<details markdown="1"><summary>Show/hide code to make palette figures</summary>
<pre><code>
```
# Wong palette
wong_pal <- c("#009E73", "#D55E00", "#F0E442",
                "#CC79A7", "#56B4E9", "#E69F00",
                "#0072B2", "#7F7F7F", "#000000")
palette(wong_pal)
gs1 <- lapply(wong_pal, function(ii, wong_pal) 
  grobTree(
    rectGrob(
      gp = gpar(fill = ii)), 
    textGrob(ii, gp = gpar(col = "white"))))
gs1 <- grid.arrange(grobs = gs1, ncol = 3, 
               top = "8-COLOR PALETTE", right = "")

# 12 color palette
twelve_pal <- c("#323232", "#BF3465", "#50B29E", "#D9D9D9", 
                 "#731683", "#1C6CCC", "#21BCFF", "#DFA5E5", 
                 "#874310", "#DB6D1B", "#B8CE17", "#F4E345")
palette(twelve_pal)
gs2 <- lapply(twelve_pal, function(ii, twelve_pal) 
  grobTree(
    rectGrob(
      gp = gpar(fill = ii)), 
    textGrob(ii, gp = gpar(col = "white"))))
gs2 <- grid.arrange(grobs = gs2, ncol = 4, 
               top = "12-COLOR PALETTE", left = "")
grid.arrange(gs1, gs2, ncol = 2)
```
</code></pre>
</details>

<br/>

```{r color_121,  layout="l-body-outset", comment='', echo=FALSE, include=FALSE}
# Wong palette
wong_pal <- c("#009E73", "#D55E00", "#F0E442",
                "#CC79A7", "#56B4E9", "#E69F00",
                "#0072B2", "#7F7F7F", "#000000")
palette(wong_pal)
gs1 <- lapply(wong_pal, function(ii, wong_pal) 
  grobTree(
    rectGrob(
      gp = gpar(fill = ii)), 
    textGrob(ii, gp = gpar(col = "white", fontsize = 10))))
gs1 <- grid.arrange(grobs = gs1, ncol = 3, 
               top = "8-COLOR PALETTE", right = "")

# 12 color palette
twelve_pal <- c("#323232", "#BF3465", "#50B29E", "#D9D9D9", 
                 "#731683", "#1C6CCC", "#21BCFF", "#DFA5E5", 
                 "#874310", "#DB6D1B", "#B8CE17", "#F4E345")
palette(twelve_pal)
gs2 <- lapply(twelve_pal, function(ii, twelve_pal) 
  grobTree(
    rectGrob(
      gp = gpar(fill = ii)), 
    textGrob(ii, gp = gpar(col = "white", fontsize = 10))))
gs2 <- grid.arrange(grobs = gs2, ncol = 4, 
               top = "12-COLOR PALETTE", left = "")
```

```{r color_8,  layout="l-body-outset", comment='', echo=FALSE, fig.height=3}
grid.arrange(gs1, gs2, ncol = 2)
```

There is a great article on [Coloring for Colorblindness](https://davidmathlogic.com/colorblind/#%23000000-%23E69F00-%2356B4E9-%23009E73-%23F0E442-%230072B2-%23D55E00-%23CC79A7) by David Nichols that has an interactive color picker and recommendations for accessible palettes. This is also a [really cool site](https://www.canva.com/colors/color-wheel/) for looking at  color combinations. Both resources are highly recommended.

</br>

<div class="post-nav">
<div class="post-nav-item">
<div class="meta-nav">Next</div>
<a href="pc.html" rel="next">Field Analyses</a>
</div>
</div>



## Source Code {.appendix}

The source code for this page can be accessed on GitHub by [clicking this link](https://github.com/hypocolypse/web/blob/master/build/intro.Rmd).