---
title: "No 5. Automated Binning"
description: |
  In this section of the workflow we use four different tools for automated binning of our assembled contigs. Our goal is to rebuild genomes from the metagenomic assembly. These are called metagenome assembled genomes, or MAGs.  
author:
  - name: Jarrod J Scott 
#    url: https://example.com/norajones
#    affiliation: Spacely Sprockets
#    affiliation_url: https://example.com/spacelysprokets
bibliography: assets/cite.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
library(kableExtra)
library(DT)
library(htmlwidgets)
library(htmltools)
```

## Automated Binning Tools

1) [CONCOCT](https://github.com/BinPro/CONCOCT) (Clustering cONtigs with COverage and ComposiTion)[@alneberg2014binning] is a program for unsupervised binning of metagenomic contigs by using nucleotide composition, coverage data in multiple samples and linkage data from paired end reads.  
2) [MetaBAT2](https://bitbucket.org/berkeleylab/metabat)[@kang2019metabat] is a statistical framework for reconstructing genomes from metagenomic data.  
3) [MaxBIN2](https://sourceforge.net/projects/maxbin2/)[@wu2016maxbin] is an automatic tool for binning metagenomics sequences.  
4) [DASTOOL](https://github.com/cmks/DAS_Tool)[@sieber2018recovery] is an automated method that integrates the results of a  number of binning algorithms to calculate an optimized, non-redundant set of bins from a single assembly.

## Run Tools

The latest iteration of anvi'o (v6 as of this writing) ports several popular automated binning algorithms into its ecosystem. We will bin the water meatgenomic assembly using  CONCOCT, MetaBAT2, and MaxBIN2. Then we use DASTOOL to integrate the results of the other binning tools. The anvio commands below will a) bin the assembly using the specified tool and b) add the results to a collection in the `PROFILE.db`. For these steps we use the command `anvi-cluster-contigs`. 

Here we run CONCOCT in two different ways. The first is using the defaults and the second is where we speciffy the number of bins. To my knowledge CONCOCT is the only one of these tools that allows you to specify the number of clusters. 

```bash
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  CONCOCT --driver concoct --just-do-it
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  CONCOCT_5 --driver concoct --just-do-it --clusters 5
```

Next the command for MetaBAT2.

```bash
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  MetaBAT2 --driver metabat2 --just-do-it --minContig 1500
```

And MaxBIN2

```bash
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  MaxBIN2 --driver maxbin2 --just-do-it --min-contig-length 1000	
```

Now we can run DASTOOL to aggregate the results. Since we ran two instances of CONCOCT we need to run both in DASTOOL against the results of MaxBIN2 and MetaBAT2. 

```bash
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  DASTOOL --driver dastool --just-do-it --search-engine diamond -S CONCOCT,MaxBIN2,MetaBAT2 --debug
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  DASTOOL_5 --driver dastool --just-do-it --search-engine diamond -S CONCOCT_5,MaxBIN2,MetaBAT2 --debug
```

<details markdown="1"><summary>Show/hide HYDRA CONTIG AUTO BINNING job script</summary>
<pre><code>
# /bin/sh
# ----------------Parameters---------------------- #
#$ -S /bin/sh
#$ -pe mthread 5
#$ -q sThC.q
#$ -l mres=25G,h_data=5G,h_vmem=5G
#$ -cwd
#$ -j y
#$ -N water_job_11_cluster_contigs
#$ -o hydra_logs/job_11_cluster_contigs_water.log
#$ -M scottjj@si.edu
#
# ----------------Modules------------------------- #
module load gcc/4.9.2
#
# ----------------Your Commands------------------- #
#
echo + `date` job $JOB_NAME started in $QUEUE with jobID=$JOB_ID on $HOSTNAME
echo + NSLOTS = $NSLOTS
#
# ----------------CALLING ANVIO------------------- #
#
export PATH=/home/scottjj/miniconda3:$PATH
export PATH=/home/scottjj/miniconda3/bin:$PATH
export PATH=/home/scottjj/miniconda3/envs:$PATH
source activate anvio-master
#
# ----------------CHECKING EVERYTHING------------------- #
#
which python
python --version
source /home/scottjj/virtual-envs/anvio-master/bin/activate
which python
python --version
which anvi-interactive 
diamond --version 
anvi-self-test -v 
#
# ----------------BINNING CONDA------------------- #
#
source activate binning
conda activate binning
which run_MaxBin.pl
which concoct
#
# ----------------SETUP TEMP DIRECTORIES------------------- #
#
rm -r /pool/genomics/stri_istmobiome/dbs/tmp_data_WATER/
mkdir -p /pool/genomics/stri_istmobiome/dbs/tmp_data_WATER/
TMPDIR="/pool/genomics/stri_istmobiome/dbs/tmp_data_WATER/"
#
# ----------------CONCOCT------------------- #
#
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  CONCOCT --driver concoct --just-do-it --debug
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  CONCOCT_5 --driver concoct --just-do-it --clusters 5 --debug
#
# ----------------MetaBAT2------------------- #
#
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  MetaBAT2 --driver metabat2 --just-do-it --debug --minContig 1500
#
# ----------------MaxBIN2------------------- #
#
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  MaxBIN2 --driver maxbin2 --just-do-it --min-contig-length 1000 --debug	
#
# ----------------DASTOOL------------------- #
#
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  DASTOOL --driver dastool --just-do-it --search-engine diamond -S CONCOCT,MaxBIN2,MetaBAT2 --debug
anvi-cluster-contigs -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -T $NSLOTS -C  DASTOOL_5 --driver dastool --just-do-it --search-engine diamond -S CONCOCT_5,MaxBIN2,MetaBAT2 --debug
#
echo = `date` job $JOB_NAME done
</code></pre>
</details>

## Inspect Results

Let's take a look at the results of the different binning strategies using the command `anvi-export-collection` with the `--list-collections` flag. Anvio won't do anything here except show us what collections are in the `PROFILE.db`.  

```bash
anvi-export-collection -p 06_MERGED/WATER/PROFILE.db --list-collections
```

And here is the output of that command. The VIRSORTER collection was imported during the annotation phase and can be ignored for now. 

```bash
COLLECTIONS FOUND
===============================================
* VIRSORTER (2840 bins, representing 2857 items).
* CONCOCT (71 bins, representing 23716 items).
* CONCOCT_5 (5 bins, representing 23716 items).
* METABAT2 (5 bins, representing 1442 items).
* MAXBIN2 (16 bins, representing 22489 items).
* DASTOOL (7 bins, representing 4775 items).
* DASTOOL_5 (2 bins, representing 956 items).
```

Here we can see the number of bins and the total number of contigs included in each collection. Remember the original assembly had 23,758 contigs (at a minimum length of 1000bp). We can take a quick look at the estimates of genome completion and redundancy using domain-specific single-copy core
genes for each collection using `anvi-estimate-genome-completeness`. 

One thing to remember is  how each algorithm names bins, or else this can get confusing. CONCOCT adds the prefix `Bin_`, MaxBin adds the prefix `MAXBIN_`, and MetaBAT adds the prefix `METABAT__` (with two underscores). DAS Tool then adds its own `Bin_` prefix to the parent name. For example DAS Tool `Bin_Bin_11` is CONCOCT bin `Bin_11` while DASTOOL `Bin_METABAT__2` is MetaBAT2 bin `METABAT__2`. 

### CONCOCT

CONCOCT bins with default parameters.

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C CONCOCT
```

Here are the results for the 71 CONCOCT bins.



```{r concoct_table, echo=FALSE, layout="l-body-outset"}
concoct <- read.table("files/concoct.txt", sep = "\t", header = TRUE)

datatable(
  concoct, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.1: CONCOCT binning results"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brltip", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 5, FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))
```

### CONCOCT 5

CONCOCT bins with the `--clusters` flag equal to 5.

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C CONCOCT_5
```

Here are the results for the 5 CONCOCT bins with the `--clusters` flag equal to 5.



```{r concoct5_table, echo=FALSE, layout="l-body-outset"}

concoct5 <- read.table("files/concoct-5.txt", sep = "\t", header = TRUE)

datatable(
  concoct5, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.2: CONCOCT 5 fored cluster binning results"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brti", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 5, FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))
```


### MetaBAT2

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C METABAT2
```
Here are the results for the 5 MetaBAT2 bins.



```{r metabat2_table, echo=FALSE, layout="l-body-outset"}
metabat2 <- read.table("files/metabat2.txt", sep = "\t", header = TRUE)

datatable(
  metabat2, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.3: MetaBAT2 binning results"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brti", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 5, FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))
```


### MaxBIN2

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C MAXBIN2
```
Here are the results for the 16 MaxBIN2 bins.

```{r maxbin2_table, echo=FALSE, layout="l-body-outset"}
maxbin2 <- read.table("files/maxbin2.txt", sep = "\t", header = TRUE)

datatable(
  maxbin2, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.4: MaxBIN2 binning results"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brltip", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 5, lengthMenu = c(5, 20), FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))

```

### DASTOOL

And finnally we can see what happens when we let DASTOOL choose bins from the combination of the other three tools. 

First the results using the default CONCOCT settings. 

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C DASTOOL
```

```{r dastool_table, echo=FALSE, layout="l-body-outset"}
dastool <- read.table("files/dastool.txt", sep = "\t", header = TRUE)

datatable(
  dastool, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.5: DASTOOL binning results"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brti", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 7, FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))
```

Here we see that DASTOOL used at least 1 bin from each of the three methods.

And then  the results using the CONCOCT `--clusters` flag equal to 5. 

```bash
anvi-estimate-genome-completeness -c 03_CONTIGS/WATER-contigs.db -p 06_MERGED/WATER/PROFILE.db -C DASTOOL_5
```

```{r dastool5_table, echo=FALSE, layout="l-body-outset"}
dastool5 <- read.table("files/dastool-5.txt", sep = "\t", header = TRUE)

datatable(
  dastool5, rownames = FALSE, autoHideNavigation = TRUE, width = "100%",  
  colnames = c(
    "bin name", "domain", "confidence", 
    "% completion", "% redundancy", 
    "num_splits", "total length"),
  caption = htmltools::tags$caption(
    style = "caption-side: bottom; text-align: left;",
    "Table 9.6: DASTOOL binning results for CONCOCT forced cluster 5"),
  extensions = "Buttons", "FixedColumns",
  options = list(
    columnDefs = list(list(className = "dt-left", targets = 0)),
    dom = "Brti", 
    buttons = c("csv", "copy", I('colvis')),
    pageLength = 7, FixedColumns = list(leftColumns = 2, rightColumns =1),
    scrollY=FALSE, scroller=TRUE))
```


However, here DASTOOL did not choose any of the CONCOCT bins. 
















<div class="post-nav">
<div class="post-nav-item">
<div class="meta-nav">Previous</div>
<a href="mg-workflow-2.html" rel="next">N<sup><u>o</u></sup> 4. Assembley & Annotation Summary</a>
</div>
</div>

<div class="post-nav">
<div class="post-nav-item">
<div class="meta-nav">Next</div>
<a href="mg-workflow-2.html" rel="prev">N<sup><u>o</u></sup> 6. SOMETHING</a>
</div>
</div>

<p class="edit-page">
  <a href="https://github.com/hypocolypse/web/blob/master/build/mg-auto-binning.Rmd">
    <i class="fas fa-pen pr-2"></i> Edit this page
  </a>
</p>
